File: java/CavanJava/src/com/cavan/java/CavanTimedArray.java
package com.cavan.java;

import java.util.ArrayList;

@SuppressWarnings("serial")
public class CavanTimedArray<E> extends ArrayList<CavanTimedNode<E>> {

	protected double mFreq;
	protected long mCycle;
	protected CavanTimedNode<E> mLastNode;

	private long mOverTime;

	public CavanTimedArray(long overtime) {
		mOverTime = overtime;
	}

	public void setOverTime(long time) {
		mOverTime = time;
	}

	public double getFreq() {
		return mFreq;
	}

	public long getCycle() {
		return mCycle;
	}

	protected CavanTimedNode<E> removeTimedNode(int index) {
		return remove(index);
	}


	public void addTimedNode(CavanTimedNode<E> node) {
		mLastNode = node;

		removeOvertimeNodes(node.getTime());
		add(node);
	}

	public void addTimedValue(E value) {
		addTimedNode(new CavanTimedNode<E>(value));
	}

	public E getTimedValue(int index) {
		return get(index).getValue();
	}

	public boolean hasTimedValue(E value) {
		for (CavanTimedNode<E> node : this) {
			if (node.getValue().equals(value)) {
				return true;
			}
		}

		return false;
	}

	protected int removeTimedNodes(long timeNow, long overtime) {
		int count = 0;

		while (true) {
			CavanTimedNode<E> first;

			try {
				first = get(0);
			} catch (IndexOutOfBoundsException e) {
				break;
			}

			if (timeNow - first.getTime() < overtime) {
				break;
			}

			removeTimedNode(0);
			count++;
		}

		return count;
	}

	protected int removeOvertimeNodes(long timeNow) {
		return removeTimedNodes(timeNow, mOverTime);
	}

	protected int removeOvertimeNodes() {
		return removeTimedNodes(System.currentTimeMillis(), mOverTime);
	}

	protected double updateFreq() {
		int size = size();
		if (size > 1) {
			int count = size - 1;
			long timeFirst = get(0).getTime();
			long timeLast = get(count).getTime();
			long timeNow = System.currentTimeMillis();

			mCycle = (timeLast - timeFirst) / count;

			if (timeNow - timeLast > mCycle) {
				mFreq = 1000.0 * size / (timeNow - timeFirst);
			} else {
				mFreq = 1000.0 / mCycle;
			}
		} else if (size < 1) {
			mFreq = 0;
		}

		return mFreq;
	}
}

File: xcode/JwaooToy/JwaooToy/CavanTimedArray.m
//
//  CavanTimedList.m
//  JwaooToy
//
//  Created by 曹福昂 on 16/8/31.
//  Copyright © 2016年 曹福昂. All rights reserved.
//

#import "CavanTimedArray.h"

@implementation CavanTimedNode

@synthesize time = mTime;

- (CavanTimedNode *)initByTime:(NSTimeInterval)time {
    if (self = [super init]) {
        mTime = time;
    }

    return self;
}

- (CavanTimedNode *)init {
    return [self initByTime:[NSDate timeIntervalSinceReferenceDate]];
}

@end

@implementation CavanTimedArray

@synthesize overtime = mOvertime;
@synthesize freq = mFreq;
@synthesize cycle = mCycle;

- (CavanTimedArray *)initByOvertime:(NSTimeInterval)overtime {
    if (self = [super init]) {
        mOvertime = overtime;
        mNodes = [NSMutableArray new];
    }

    return self;
}

- (void)removeTimedNode:(NSInteger)index {
    [mNodes removeObjectAtIndex:index];
}

- (CavanTimedNode *)getFirstNode {
    return mNodes.firstObject;
}

- (CavanTimedNode *)getLastNode {
    return mNodes.lastObject;
}

- (NSTimeInterval)getFirstTime {
    CavanTimedNode *node = self.getFirstNode;
    if (node != nil) {
        return node.time;
    }

    return 0;
}

- (NSTimeInterval)getLastTime {
    CavanTimedNode *node = self.getLastNode;
    if (node != nil) {
        return node.time;
    }

    return 0;
}

- (NSTimeInterval)getTimeInterval {
    return self.getLastTime - self.getFirstTime;
}

- (NSUInteger)removeOvertimeNodes:(NSTimeInterval)timeLast {
    NSUInteger count = 0;

    while (1) {
        CavanTimedNode *firstNode = mNodes.firstObject;
        if (firstNode == nil || firstNode.time > timeLast) {
            break;
        }

        [self removeTimedNode:0];
        count++;
    }

    return count;
}

- (NSUInteger)removeOvertimeNodes {
    NSTimeInterval timeLast = [NSDate timeIntervalSinceReferenceDate] - mOvertime;
    return [self removeOvertimeNodes:timeLast];
}

- (void)addTimedNode:(CavanTimedNode *)node {
    NSTimeInterval timeLast = node.time - mOvertime;
    [self removeOvertimeNodes:timeLast];
    [mNodes addObject:node];
}

- (double)updateFreq {
    NSUInteger count = mNodes.count;
    if (count > 1) {
        CavanTimedNode *firstNode = mNodes.firstObject;
        CavanTimedNode *lastNode = mNodes.lastObject;

        mCycle = (lastNode.time - firstNode.time) / (count - 1);

        NSTimeInterval timeNow = [NSDate timeIntervalSinceReferenceDate];
        NSTimeInterval interval = timeNow - lastNode.time;
        if (interval < mCycle) {
            mFreq = 1.0 / mCycle;
        } else {
            mFreq = count / (timeNow - firstNode.time);
        }
    } else if (count < 1) {
        mFreq = 0;
    }

    return mFreq;
}

@end
